### PR TITLE
[inductor] no-side-effect codegen

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108193
* __->__ #107617

Inductor kernel codegen previously have the following side effect:
- in `Kernel.__exit__ `, we add local used buffers in graph.removed_buffers
- during codegen, we do memory allocation/free.

These cause doing multiple versions of codegen for the same kernel hard. The PR refactor the code to make kernel codegen not changing graph level states. After codegening a kernel, the graph level state is not changed so we can go on to codegen another version of the kernel if we want.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -2379,10 +2379,18 @@ def codegen_node_schedule(self, node_schedule, numel, reduction_numel):
 
         self.codegen_node_schedule_with_kernel(node_schedule, kernel)
 
-        src_code = kernel.codegen_kernel()
+        with V.set_kernel_handler(kernel):
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:2382]: could we combine this with the set kernel handler line below?

**Review by jansel — CHANGES_REQUESTED** [review]
tests are failing

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:2382]: yea, merged locally. Will send out a update after I look into the test failures.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**